### ISSUE TITLE
torch.cat different Windows vs Linux behavior 

### ISSUE THREAD

**Issue by morawi** [pytorch/pytorch#52233]
## 🐛 Bug

The below statement ran well under Windows (10), but caused a type mismatch under Linux. It is a simple and clear problem.

## To Reproduce
`xx =torch.cat(( torch.tensor([]),   torch.tensor([1,2,3], dtype=torch.long)))
`
as` torch.tensor([]) `is float32, the second term is casted to float32  Windows, in Windows, we will have:
```
torch.typename(xx)
'torch.FloatTensor'

```
but Linux it threw an error message: 
RuntimeError: Expected object of scalar type Float but got scalar type Long for sequence element 1 in sequence argument at position #1 'tensors'

I'm sorry to be lazy, but I think the problem is clear and will thus skip a few reporting issues shown below.  

1.
1.
1.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): 1.7.1
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration: 
 - Any other relevant information:

## Additional context

cc @nairbv @mruberry

**Comment by ngimel** [comment]
Can't repro on master

**Comment by morawi** [comment]
> Can't repro on master
I'm using Python Python 3.7.9 (in Windows) in my Laptop (local machine) ... and Linux on the HPC cluster. 
Seems I'm using PyTorch 1.4 on Linux and 1.7 on Windows. Probably it is due to torch 1.4? Apologies if so. 
I've just tried it again:

```
$ python
Python 3.7.4 (default, Aug 13 2019, 20:35:49)
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.__version__
'1.4.0'
>>> torch.cat(( torch.tensor([]), torch.tensor([1,2,3], dtype=torch.long)))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Expected object of scalar type Float but got scalar type Long for sequence element 1 in sequence argument at position #1 'tensors'

```
This works fine: 

```
>>> torch.cat(( torch.tensor([]), torch.tensor([1,2,3], dtype=torch.float)))
tensor([1., 2., 3.])

```
OS used 
```
$ egrep '^(VERSION|NAME)=' /etc/os-release
NAME="Scientific Linux"
VERSION="7.8 (Nitrogen)"

```


Now, Windows-10 Pro (Version 10.0.18363 Build 18363):

```
torch.__version__
Out[16]: '1.7.1'

torch.cat(( torch.tensor([]), torch.tensor([1,2,3], dtype=torch.long)))
Out[22]: tensor([1., 2., 3.])

```

**Comment by nairbv** [comment]
> >>> torch.__version__
> '1.4.0'
> >>> torch.cat(( torch.tensor([]), torch.tensor([1,2,3], dtype=torch.long)))
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> RuntimeError: Expected object of scalar type Float but got scalar type Long for sequence element 1 in sequence 

This may have been fixed in v1.6.0:
https://github.com/pytorch/pytorch/releases/tag/v1.6.0

> torch.cat: fixed missing type promotion (#35030, #39777)

I suspect this works in windows because you're using PyTorch 1.7.1 there, not because of the OS difference.

**Comment by gchanan** [comment]
going to close given this seems to be a version difference issue.  Please reopen if you disagree.
